### PR TITLE
SDK-1139. Improve folder upload/download controllers (Merge recursive approach)

### PR DESCRIPTION
**Pull request by jgandres**

### CODE REVIEW

**Comment by jgandres** [comment]
trigger compilation

**Comment by jgandres** [comment]
trigger compilation

File: include/megaapi_impl.h [hunk 1]
```diff
@@ -227,27 +227,66 @@ class MegaRecursiveOperation
...
+
+    // worker thread
+    std::thread mWorkerThread;
+
+    // thread id of main thread
+    std::thread::id mMainThreadId;
 };
 
+class TransferQueue;
 class MegaFolderUploadController : public MegaRequestListener, public MegaTransferListener, public MegaRecursiveOperation
```
Comment (mattw-mega) [inline include/megaapi_impl.h:240]: I think we can remove the inheritance from MegaRequestListener now - we no longer override any of its functions.

File: include/megaapi_impl.h [hunk 2]
```diff
@@ -393,21 +432,34 @@ class MegaBackupController : public MegaBackup, public MegaRequestListener, publ
     void setValid(bool value);
 };
 
-class MegaFolderDownloadController : public MegaTransferListener, public MegaRecursiveOperation
+class MegaFolderDownloadController : public MegaRequestListener, public MegaTransferListener, public MegaRecursiveOperation
```
Comment (mattw-mega) [inline include/megaapi_impl.h:435]: Do we need to inherit from MegaRequestListener?  We're not overriding any onRequest function so this is probably left over from earlier changes.

**Comment by jgandres** [comment]
trigger compilation

**Comment by jgandres** [comment]
trigger compilation

**Comment by jgandres** [comment]
trigger compilation

**Pull request closed without merging**